### ISSUE TITLE
Implement a cross-engine WHATWG set of shims

### ISSUE THREAD

**Issue by jviotti** [sourcemeta/includejs#22]

**Comment by tony-go** [comment]
Should this be directly in the example?

If we are a library that helps people write their runtime, I don't know if we should add more support for these APIs.

Here are the two solutions that I see:

- We let includes empty as we could, then we create another repo, a real-runtime that could be the proof of includejs usage
- Or we could create standalone modules that users can opt-in `whatwg/console`, `whatwg/timers`

**Comment by tony-go** [comment]
Ideas? @jviotti @thundron @bavulapati @RaisinTen

**Comment by RaisinTen** [comment]
Most of the WHATWG APIs are expected to be available in pretty much all JS runtimes, so I think it makes sense to implement those here. I can see how this can get tricky with stuff like `timers`. If we include it here, this would get tied with an event loop because having timers would mean running async code. An implementation that is not too coupled with an event loop would expose an interface that would need to be satisfied by whatever event loop the user wants to use.

**Comment by tony-go** [comment]
yeah, it makes sense, but I wonder if someone wants to do its own implementation of timers for whatever reason, could we have a way to opt out?

**Comment by RaisinTen** [comment]
Yea, I think it makes total sense to allow customizations.

**Comment by tony-go** [comment]
> An implementation that is not too coupled with an event loop would expose an interface that would need to be satisfied by whatever event loop the user wants to use.

Really like this :)

**Comment by robertgzr** [comment]
> expose an interface that would need to be satisfied by whatever event loop the user wants to use

sounds like we probably want to define that interface as part of includejs and provide the mvp interface on top of appkit, win32 and glib?
something like `dispatch(callback)` would probably be sufficient for a start...

the only thing that is not clear to me here is how the integrating side would hook such an intermediate into it's own event loop. for the big toolkits mentioned above it might not be a hard requirement, but if an app drives it's own event loop with libuv for example they would have to be able to implement our dispatch interface or (ideally) plug in a libuv implementation we provide
